### ISSUE TITLE
how to enable UAT/dump978?

### ISSUE THREAD

**Issue by dlasher** [tsunghanliu/adsb-box.snap#15]
Appears most of the programs are installed to support it, quick guidelines on how to enable dump978, assuming another RTL-SDR stick is already plugged in, cabled, and ready to go?

**Comment by tsunghanliu** [comment]
To support the UAT, I think that the works will be:
* include dump978
* add corresponding start wrapper
* add a new tab to the web portal

I may not have enough time to implement this feature in the near future and I'm not able to verify it because UAT isn't used in my country.
Besides US, any other country is using UAT?

**Comment by dlasher** [comment]
Agree with the outline, makes sense.

I have it running on the previous feeder box I was using, it's fairly simple, the complicated part is how to bring the data into your existing setup, either as a separate map/feed or integrated.

From what I can see, dump1090-fa has support for dump978, just not on the same process, which would be ideal given the work you've already done. Hard to really tell, since the same docs refer to installing dump978-fa.

notes:
https://flightaware.com/adsb/piaware/install
https://flightaware.com/adsb/piaware/advanced_configuration#uatsteps
https://github.com/flightaware/dump978
https://github.com/mutability/dump978

https://www.adsbexchange.com/how-to-feed/adding-978-mhz-ads-b-capability-to-your-pipi2/
^^ the method I'm using right now

Love what you've done using snaps so far, hoping we can keep adding features.. ACARS anyone? :)


https://www.sysrun.io/2015/11/16/rpi-docker-acars-receive-aircraft-sms/

**Comment by tsunghanliu** [comment]
Thanks for the information.

It's not that difficult thing. I'll find some time slots to plan and implement the UAT support.

**Comment by dlasher** [comment]
It may be easier than we think - I grabbed the 978 stuff from the flightware report (https://github.com/flightaware/dump978) and it compiles and installs cleanly. If you could get it into an Edge snap, I'll test/configure it, and let you know what it took to make it work.

**Comment by dlasher** [comment]
Especially right now, with the world in an odd state, it would be great to see a few more flights.. any chance we'll get this?

https://discussions.flightaware.com/t/how-to-add-on-978-mhz-dump978-fa-etc-with-buster-and-piaware-3-8/59737/13

**Comment by tsunghanliu** [comment]
Hi,
The UAT support is on my TO-DO list, but it requires several updates, at least:
* add dump978
* add rtl-sdr device selection to both dump1090 and dump978
* add a new tab to the web UI
* update performance graphs

I don't have too much progress yet and will let you know if there's a testing or preview version.

**Comment by dlasher** [comment]
Thanks, that would be great. Love the snap, using it several places. :)

**Comment by dlasher** [comment]
Any update on this wishlist item? Flightaware5 now includes 978 support

**Comment by tsunghanliu** [comment]
I don't have too much progress on this.
Like you said that Flightaware 5 includes dump978 support, so I will revisit this soon.

**Comment by dlasher** [comment]
Thanks, looking forward to it.

**Comment by tsunghanliu** [comment]
Hey @dlasher and folks,

I have a testing version for supporting UAT/dump978. I can only verify that the dump978 is running and new UI showed, but cannot see any UAT traffic because my country doesn't use UAT. That will be great if you could help to test.

Suppose your feeder has or will have two RTL-SDR dongles to collect both ADS-B and UAT signals.

## highlight changes
* upgrade to ubuntu core 20
* add dump978
* use [graph1090](https://github.com/wiedehopf/graphs1090) instead of ADS-B receiver project
* upgrade rbfeeder to the latest version (to feed UAT data)

## refresh to testing version

```bash
# stop adsb-box temporarily
$ sudo snap stop adsb-box 
# backup your current settings and data
$ sudo snap save adsb-box
# refresh
$ sudo snap refresh adsb-box --channel edge/uat978

# here it takes about 8 minutes to migrate the RRD DB of graphs
# please be patient
$ sleep 480

# stop adsb-box temporarily again
$ sudo snap stop adsb-box 
```

## provision dongle
This step is used to program the serial numbers in the EEPROM of dongles. The serial numbers will be provided to `dump1090` and `dump978` later.
It would be easier to provision dongles one by one. If you find that the serial (SN) is not `00000001`, you can skip this step.

```bash
# run rtl-sdr tools within the snap
$ sudo snap run --shell adsb-box.dump1090
(in-snap)$ cd $SNAP
# list dongles
(in-snap)$ rtl_test -t
Found 1 device(s):
  0:  Realtek, RTL2838UHIDIR, SN: 00000001
  ...
# set serial number to a 8 digit number, eg 10900001 or 09780001
# -d for index, -s for serial
(in-snap)$ rtl_eeprom -d 0 -s 10900001 

# unplug the dongle and plug another one
# re-do rtl_test and rtl_eeprom commands
# remember to use a different serial for the second dongle

(in-snap)$ exit
``` 

##  enable dump978
Replace serials with yours set in the previous step.
In this example, dongle 10900001 is for dump1090 and dongle 09780001 is for dump978.

```bash
$ sudo snap set adsb-box receiver.dump1090.serial=10900001
$ sudo snap set adsb-box receiver.dump978.serial=09780001
$ sudo adsb-box.piaware-config uat-receiver-type sdr
```

## restart adsb-box
Connect both dongles and their cables and antennas.

```bash
$ sudo snap restart adsb-box
```

## verify
- [ ] On the skyview page, you would see `UAT` in the legend list.
- [ ] On the skyview page, you would see aircraft using UAT.
- [ ] On the performance view, you would see a new group of dump978 charts.
- [ ] Check your feeder page on Flightaware if it recevies your UAT data.
- [ ] Check your feeder page on Radarbox24 if it receives you UAT data.

Please let me know if you have any questions or suggestions.
Thank you.

**Comment by tsunghanliu** [comment]
arm64 version is still waiting for the Launchpad builder to build. I'll publish it later.

i386 is not supported in Ubuntu 20.04 and Core 20. So, i386 platform will stick on v1.5.x.

**Comment by dlasher** [comment]
Awesome, thank you!

Walked through the changes, everything "seems" to be working, but no 978 data yet.. will let it run a few hours and see.. I only get a few 978 a day, mostly traffic from the local flight school, so I'll give it time.

Linux pi-adsb 4.15.0-1084-raspi2 #89-Ubuntu SMP PREEMPT Tue Apr 13 12:52:44 UTC 2021 armv7l armv7l armv7l GNU/Linux

I do see 978 processes running:

_root      2488  0.0  0.2   4344  2328 ?        Ss   09:12   0:00 /bin/bash /snap/adsb-box/428/bin/skyaware978
root      2560  0.1  0.3  45200  2952 ?        Sl   09:12   0:00 /snap/adsb-box/428/usr/bin/skyaware978-fa --connect localhost:30978 --reconnect-interval 30 --history-count 120 --history-interval 30 --json-dir /tmp/skyaware978 --lat XX.XXX --lon XX.XXX_

Setting receiver type did complain a bit, putting in here in case it's important.

_root@pi-adsb:~# adsb-box.piaware-config uat-receiver-type sdr
+ ARCH_TRIPLET=arm-linux-gnueabihf
+ find /snap/adsb-box/428/usr/share/tcltk/ -maxdepth 1 -type d -name tcl8*
+ DIR_TCL8=/snap/adsb-box/428/usr/share/tcltk/tcl8.6
+ find /snap/adsb-box/428/usr/share/tcltk/ -maxdepth 1 -type d -name tcllib*
+ DIR_TCLLIB=/snap/adsb-box/428/usr/share/tcltk/tcllib1.20
+ find /snap/adsb-box/428/usr/share/tcltk/ -maxdepth 1 -type d -name itcl*
+ DIR_ITCL=/snap/adsb-box/428/usr/share/tcltk/itcl3.4
+ TCL_LIBRARY=/snap/adsb-box/428/usr/share/tcltk/tcl8.6:
+ TCLLIBPATH= /snap/adsb-box/428/usr/share/tcltk/tcl8.6
+ find /snap/adsb-box/428/usr/share/tcltk -type d -maxdepth 1 -mindepth 1
find: warning: you have specified the global option -maxdepth after the argument -type, but global options are not positional, i.e., -maxdepth affects tests specified before it as well as those specified after it.  Please specify global options before other arguments.
find: warning: you have specified the global option -mindepth after the argument -type, but global options are not positional, i.e., -mindepth affects tests specified before it as well as those specified after it.  Please specify global options before other arguments.
+ TCLLIBPATH= /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/itcl3.4
+ TCLLIBPATH= /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/itcl3.4 /snap/adsb-box/428/usr/share/tcltk/tcl8.6
+ TCLLIBPATH= /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/itcl3.4 /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/tcllib1.20
+ find /snap/adsb-box/428/usr/lib/tcltk/arm-linux-gnueabihf -type d -maxdepth 1 -mindepth 1
find: warning: you have specified the global option -maxdepth after the argument -type, but global options are not positional, i.e., -maxdepth affects tests specified before it as well as those specified after it.  Please specify global options before other arguments.
find: warning: you have specified the global option -mindepth after the argument -type, but global options are not positional, i.e., -mindepth affects tests specified before it as well as those specified after it.  Please specify global options before other arguments.
+ TCLLIBPATH= /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/itcl3.4 /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/tcllib1.20 /snap/adsb-box/428/usr/lib/tcltk/arm-linux-gnueabihf/itcl3.4
+ TCLLIBPATH= /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/itcl3.4 /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/tcllib1.20 /snap/adsb-box/428/usr/lib/tcltk/arm-linux-gnueabihf/itcl3.4 /snap/adsb-box/428/usr/lib/tcltk/arm-linux-gnueabihf/tcltls1.7.20
+ ITCL_LIBRARY=/snap/adsb-box/428/usr/share/tcltk/itcl3.4
+ export TCL8_6_TM_PATH=/snap/adsb-box/428/usr/share/tcltk/tcl8.6/tcl8
+ export TCL_LIBRARY=/snap/adsb-box/428/usr/share/tcltk/tcl8.6:
+ export TCLLIBPATH= /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/itcl3.4 /snap/adsb-box/428/usr/share/tcltk/tcl8.6 /snap/adsb-box/428/usr/share/tcltk/tcllib1.20 /snap/adsb-box/428/usr/lib/tcltk/arm-linux-gnueabihf/itcl3.4 /snap/adsb-box/428/usr/lib/tcltk/arm-linux-gnueabihf/tcltls1.7.20
+ export ITCL_LIBRARY=/snap/adsb-box/428/usr/share/tcltk/itcl3.4
+ /snap/adsb-box/428/usr/bin/piaware-config uat-receiver-type sdr
warning: /var/snap/adsb-box/428/piaware/piaware.conf:1: unrecognized option flightaware-user
warning: /var/snap/adsb-box/428/piaware/piaware.conf:2: unrecognized option flightaware-password_

**Comment by dlasher** [comment]
I don't see dump978-fa running however, suspect that's why I have no data.

**Comment by dlasher** [comment]
SoapySDR seems unhappy. 

root@pi-adsb:~# snap run --shell adsb-box.dump1090
root@pi-adsb:/root# dump978-fa --sdr driver=rtlsdr,serial=00000978 --raw-port 30978 --json-port 30979       
raw-port: listening for connections on 0.0.0.0:30978
raw-port: listening for connections on [::]:30978
json-port: listening for connections on 0.0.0.0:30979
json-port: listening for connections on [::]:30979
Configuration error: No matching SoapySDR device found (cause: SoapySDR::Device::make() no match)

That is the right device.. mine are:
00000978
00001090

thoughts?

**Comment by tsunghanliu** [comment]
Thanks @dlasher 

I made a mistake in dump978 wrapper. Please refresh to ver v1.5.1-38-g4fdeddf. It should fix the issue of dump978-fa.

**Comment by tsunghanliu** [comment]
> arm64 version is still waiting for the Launchpad builder to build. I'll publish it later.

arm64 version is available now.

**Comment by dlasher** [comment]
Looking good so far!

root     26389  0.1  0.2   4344  2324 ?        Ss   19:34   0:00 /bin/bash /snap/adsb-box/432/bin/dump978
root     26425 51.1  1.1  73228 10864 ?        Sl   19:34   0:31 /snap/adsb-box/432/usr/bin/dump978-fa --sdr driver=rtlsdr,serial=00000978 --format CS8 --raw-port localhost:30978 --json-port localhost:30979
root     26605  0.4  0.2   4344  2320 ?        Ss   19:34   0:00 /bin/bash /snap/adsb-box/432/bin/skyaware978
root     26691  0.1  0.3  45200  2848 ?        Sl   19:34   0:00 /snap/adsb-box/432/usr/bin/skyaware978-fa --connect localhost:30978 --reconnect-interval 30 --history-count 120 --history-interval 30 --json-dir /tmp/skyaware978 --lat 45.689017 --lon -122.629534
root     27065  0.1  0.3  45036  2952 ?        Sl   19:34   0:00 /snap/adsb-box/432/usr/lib/piaware/helpers/faup978 --connect localhost:30978

The right services seem to be running, and the performance/graphs exist for 978 now.

Tomorrow when the flight school puts planes up again, I should have some data. 

Thanks for the quick turn-around!

**Comment by dlasher** [comment]
![image](https://user-images.githubusercontent.com/937979/120762026-ac6a0800-c4ca-11eb-9448-abd51f0ba6a6.png)

We have UAT!

**Comment by dlasher** [comment]
And I'm tracking planes that are ONLY on 978 and not 1080.. The way I read the US requirements for 2020, all planes had to be announcing on 1080. interesting.

![image](https://user-images.githubusercontent.com/937979/120860920-4f5a6a80-c53b-11eb-9cfb-576821185a6d.png)

**Comment by tsunghanliu** [comment]
I've uploaded the code and released a version to the edge channel.
I'll promote it to candidate few days later

Note: i386 platform will not get this update because some packages are not available on Ubuntu 16.04.
